chore: add Financial Contributions section to GOVERNANCE

arendjr opened the pull request:
## Summary

This is my first attempt at adding a section for Financial Contributions to the GOVERNANCE document. Feedback very much welcome!

## Test plan

Members of the @biomejs/core-contributors need to approve the changes

Conaclos commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,81 @@ We have a reasonably liberal approach to code review and merging. We value quick
+Compared to sponsorship, bounties come with a few risks, both for Biome as a project, as well as the persons or entities offering and collecting the bounty. To minimize these risks, we only sanction bounties under the following conditions:
+
+- Bounties must be pledged through Polar.sh on an issue that is opened in our issue tracker.
+  - Unless otherwise specified, Biome reserves 30% of the payment amount, meaning contributors receive 70% of the pledged amount.
```
Do you know if 70/30 is the usual amount for bounties?

arendjr replied:
To be honest, I don't. I would be leaning to either 70/30 or 80/20, but it's hard to say which should be preferred. And the range as a whole is more a gut-feeling than anything.

Conaclos commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,81 @@ We have a reasonably liberal approach to code review and merging. We value quick
+  - The waiting period is not necessary if a lead approves the request.
+- There is a maximum amount of issues with pledges someone can have assigned to them. The limit is 2 for core contributors, and 1 for anyone else.
+
+Finally, there is one more important distinction from community-funded bounties: Since Biome is the one funding the issue, 100% of the pledged amount will go to the contributor completing the task.
```
```suggestion
Finally, there is one more important distinction from community-funded bounties: Since Biome is the one funding the issue, **100% of the pledged amount will go to the contributor completing the task**.
```

Conaclos commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,81 @@ We have a reasonably liberal approach to code review and merging. We value quick
+
+### Project-Funded Bounties
+
+In some cases, Biome funds bounties on its own issues. For project-funded bounties, the same conditions apply as for community-funded bounties. But given that these issues are funded by money from our sponsors, we have additional restrictions in place:
```
Do we also use polar.sh? Maybe we could directly ask the contributor to retrieve the money from the Open Collective? This should avoid some fees?

arendjr replied:
Yes, that's a good point!

Conaclos reviewed (commented):
Looks good at me :) I left some suggestions.

arendjr commented:
Added two more rules:

- A task is only completed when a Biome maintainer merges the pull request that closes the task. The person trying to complete the task is responsible for addressing review comments to make sure the work meets the quality standards of the Biome project.
- (At @Conaclos request:) For Biome-funded tasks assigned to core contributors, we ask the contributor to take the money directly from our Open Collective account, instead of using Polar.sh. This will help us avoid paying commission over the larger bounties.

ematipico commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,84 @@ We have a reasonably liberal approach to code review and merging. We value quick
+
+### Project-Funded Bounties
+
+In some cases, Biome funds bounties on its own issues. For project-funded bounties, the same conditions apply as for community-funded bounties. But given that these issues are funded by money from our sponsors, we have additional restrictions in place:
```
> But given that these issues are funded by money from our sponsors, we have additional restrictions in place:

This isn't clear to me. Sponsors don't fund a "bounty", they fund Biome as a whole with some perks, but they fund particular "projects" or "issues". Maybe we should revisit this phrase.

Also, how does Biome funds itself? This needs to be stated, because this isn't clear to me.

arendjr replied:
All our money comes from sponsors, so any time we fund issues through bounties, we're spending money received from our sponsors. I'll see if I can adjust the wording a bit.

ematipico replied:
Actually, no. It isn't possible (nowadays) to fund issues via polar.sh using money coming from our open collective.

arendjr replied:
Ah, that’s an important caveat indeed. Then I’ll just revise this to remain limited to core contributions for now.

arendjr replied:
Done.

ematipico commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,84 @@ We have a reasonably liberal approach to code review and merging. We value quick
+
+In some cases, Biome funds bounties on its own issues. For project-funded bounties, the same conditions apply as for community-funded bounties. But given that these issues are funded by money from our sponsors, we have additional restrictions in place:
+
+- Implementation of the task funded by Biome must advance [our roadmap](#roadmap) or help with essential infrastructure.
+  - The description of the task must include a clear rationale for why this task is eligible for funding.
+- Each Biome-funded task will be either assigned directly to a contributor, or assigned a label to indicate who is eligible for picking up that task:
+  - `F-Anyone`: Anyone is free to pick up this task. This task should be of low complexity and has a maximum bounty of $50 (USD).
+  - `F-Maintainers`: Any Biome maintainer is free to pick up this task. This task may be of medium complexity and has a maximum bounty of $200 (USD).
+  - `F-Core-Contributors`: Only core contributors are eligible for picking this task. There is no upper limit to the complexity, but the maximum bounty is $1000 (USD).
+- For Biome-funded tasks assigned to core contributors, we ask the contributor to take the money directly from our Open Collective account, instead of using Polar.sh. This will help us avoid paying commission over the larger bounties.
```
I see some confusion here, maybe because it isn't stated precisely how these funds are allocated and withdrawn.

This gets more confusing after reading the core contributor section, where they should withdraw money from the open collective.

So, here's the main question: what's the primary intention of these "Project-Funded Bounties"? How do they different from the previous funding system?

arendjr replied:
I've updated this section. Hopefully this question is resolved.

ematipico commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,84 @@ We have a reasonably liberal approach to code review and merging. We value quick
+
+- To request Biome funding for a task, a request with a clear rationale must be made in the `Community > #bounties` channel on Discord.
+  - Merely commenting on a GitHub issue is not sufficient, to make sure project members have a better chance at monitoring requests.
+- If the request is approved by at least 2 maintainers and not vetoed by any core contributor within 7 days, then a core contributor can assign the pledge to the issue and assign the requester or a corresponding label to it.
```
The community-funded bounties were decided by the core contributors, but these "bounties" are decided by maintainers too. Is this intentional or a mistake? This discrepancy should be clarified.

Also, I don't think Maintainers should be entitled to decide on funding matters, only the core team should be.

arendjr replied:
Good catch, that's indeed a mistake. I'll just simplify it to "If no Core contributor raises objections within 7 days" and drop the part about maintainer approval.

ematipico commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,84 @@ We have a reasonably liberal approach to code review and merging. We value quick
+Additionally, there is a strict process for assigning project-funded bounties:
+
+- To request Biome funding for a task, a request with a clear rationale must be made in the `Community > #bounties` channel on Discord.
+  - Merely commenting on a GitHub issue is not sufficient, to make sure project members have a better chance at monitoring requests.
```
```suggestion
  - Commenting on a GitHub issue might not guarantee enough visibility to ensure project members have a better chance at monitoring requests.
```

I don't know what this bullet point means; I rephrase it in a way where commenting is still a legitimate approach, but it doesn't guarantee the desired outcome.

arendjr replied:
The point here is that Core contributors should only need to watch a single channel without needing to worry they might miss funding-related decisions being made outside that channel. I think this is less of a concern for community-funded issues, since it isn't Biome's money being spent, but it's a relevant concern for issues funded by Biome itself.

So I don't agree with the suggestion here, because it leaves open the possibility that such decisions get made in an issue thread instead. But I guess I may need to reword a bit to make the intention more clear.

ematipico commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,84 @@ We have a reasonably liberal approach to code review and merging. We value quick
+  - Merely commenting on a GitHub issue is not sufficient, to make sure project members have a better chance at monitoring requests.
+- If the request is approved by at least 2 maintainers and not vetoed by any core contributor within 7 days, then a core contributor can assign the pledge to the issue and assign the requester or a corresponding label to it.
+  - The waiting period is not necessary if a lead approves the request.
+- There is a maximum amount of issues with pledges someone can have assigned to them. The limit is 2 for core contributors, and 1 for anyone else.
```
Is there a specific reason for this limit? I agree with this point, however should explain it

arendjr replied:
Mainly that you can't be working on too many issues at once anyway. But by having them assigned to you, you deprive others of the opportunity to work on them.

arendjr replied:
Added a small explainer.

ematipico commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,84 @@ We have a reasonably liberal approach to code review and merging. We value quick
+  - The waiting period is not necessary if a lead approves the request.
+- There is a maximum amount of issues with pledges someone can have assigned to them. The limit is 2 for core contributors, and 1 for anyone else.
+
+Finally, there is one more important distinction from community-funded bounties: Since Biome is the one funding the issue, **100% of the pledged amount will go to the contributor completing the task**.
```
This is a very important point, and I think should at the very beginning 😅

Conaclos replied:
Agreed, I was wondering after the first paragraph if that wasn't missing 😅.

ematipico reviewed (commented):
I left some suggestions and questions. There's some opacity around these "Project-funded projects", I didn't understand what they are.

arendjr commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,84 @@ We have a reasonably liberal approach to code review and merging. We value quick
+  - Implementation of the task is (currently) technically infeasible.
+  - The suggested pledge is deemed inappropriate for the amount of work involved.
+- The person making the pledge acknowledges their pledge does not come with a guarantee of work done.
+- A task is only completed when a Biome maintainer merges the pull request that closes the task. The person trying to complete the task is responsible for addressing review comments to make sure the work meets the quality standards of the Biome project.
```
It was intentional, since maintainers have merge permissions. But if it makes more sense, we can of course include a rule to specify only Core contributors should merge PRs that are for bounties.

ematipico replied:
No, I think it's fine :)

arendjr commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,84 @@ We have a reasonably liberal approach to code review and merging. We value quick
+
+### Community-Funded Bounties
+
+Bounties are the second approach towards funding Biome development. Unlike a sponsorship, bounties are directed towards the completion of a specific task. Tasks include, but are not limited to, the development of a feature or the fixing of a bug.
```
So if someone wants to adopt Biome but finds there's an issue blocking them, I think it's very natural to want to fund said issue. I'm not really sure I understand the objection? Of course we may wish for them to fix bugs themselves, but we have to recognize not everyone has the skills or time to do so. If anything, I'd say financial contributors are part of the community too :)

There's also another reason why I'm hesitant towards making a distinction that would disallow bounties on bug fixes: Developers and users regularly argue about what should be considered a bug and what is a (missing) feature. By saying one can be funded and the other cannot, we even create a financial incentive for these arguments to take place. I think it's not a good idea to make such a distinction.

ematipico replied:
This could incentivize behaviours where  people would fix bugs only if paid, which is something I am really afraid of and I wouldn't want to see.

If someone wants a bug fixed ASAP, I would rather see them sponsor the project so we could afford someone working part/full time on maintenance or bug fixes.

I've seen people implementing features (in Biome) and then disappearing, and **I** had to fix the bugs.

It's not about the users that want their bugs fixed, but it's about having the right people and resources to fix a bug.

arendjr replied:
Doc is updated :)

arendjr commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,84 @@ We have a reasonably liberal approach to code review and merging. We value quick
+To learn more about sponsorship incentives, or to make a donation, please visit our [Open Collective page](https://opencollective.com/biome).
+You can also sponsor the project through [GitHub](https://github.com/sponsors/biomejs).
+
+### Community-Funded Bounties
```
I'm not sure if there's a better alternative, but the words "issue" and "project" are already used for units of work, which may or may not be funded. The word "bounty" unambiguously clarifies there's a financial reward for a unit of work. The Polar terminology makes sense within their realm, since the platform is all about funding, so they differentiate "issues" from other financial transactions, but applying that terminology to Biome would be more confusing than helpful, I think.

I don't think there's an implication that bounties should only be for small things. That should be reflected in the pledged amount instead. But I do think issues with a bounty attached shouldn't become *too* big, hence the $1000 maximum. If a task is so large we feel $1000 doesn't cover the effort, it would be better to split it into multiple tasks with separate bounties.

Conaclos replied:
Community-funded issue?

arendjr replied:
I can change the header, yes. But the word bounty is used all-over the place in this doc, and using the full phrase "Community-funded issue" everywhere seems quite burdensome. Assuming we don't want to ban the word bounty altogether,  I'm not a fan of changing the header either, since it would only create a disconnect between the terminology used.

arendjr commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,84 @@ We have a reasonably liberal approach to code review and merging. We value quick
+- Bounties must be pledged through Polar.sh on an issue that is opened in our issue tracker.
+  - Unless otherwise specified, Biome reserves 30% of the payment amount, meaning contributors receive 70% of the pledged amount.
+  - The commission charged by Polar.sh is split accordingly.
+- If an issue in our issue tracker is not yet open for pledges, the person who wishes to pledge should indicate their desire in the comments of the issue, or do so through the `Community > #bounties` channel on Discord. A core contributor can then **accept** or **reject** the request for the bounty. Reasons for rejection include, but are not limited to: 
```
All done, except the last point: I wonder if/hope it might be sufficient for now to let core contributors make these decisions at their own discretion, as long as it fits within the rest of the rules as stated in this document. Do we think a poll is necessary for each bounty request?

ematipico commented on GOVERNANCE.md:
```diff
@@ -165,6 +174,80 @@ We have a reasonably liberal approach to code review and merging. We value quick
+
+### Project-Funded Bounties
+
+In some cases, Biome funds bounties on its own issues in order to further the project goals as laid out in its [roadmap](#roadmap). Because Biome is the one funding these issues, it also means **100% of the pledged amount will go to the contributor completing the task**. Project-Funded Bounties have the same conditions applied to them as [Community-Funded Bounties](#community-funded-bounties). Bu ...
+
+- Implementation of the task funded by Biome must advance our [roadmap](#roadmap) or help with essential infrastructure.
+  - The description of the task must include a clear rationale for why this task is eligible for funding.
+- To avoid paying commission on our own bounties, we do not assign funds through Polar.sh. Instead, we ask contributors to take the money directly from our Open Collective account upon completion of the task.
+  - Unfortunately, this means Project-Funded Bounties can only be assigned to [Core Contributors](#core-contributor) at this moment, since they have access to our Open Collective account.
+
+Additionally, there is a strict process for assigning Project-Funded Bounties:
+
+- To request Biome funding for a task, a request with a clear rationale must be made in the `Community > #funding` channel on Discord.
+  - To make sure our Core Contributors don't have to worry about overlooking Biome-funding-related decisions, requests made on GitHub issues are not sufficient for a funding request.
+- If no Core Contributor raises objections within 7 days, then any Core Contributor can assign the pledge to the issue and assign the requester to it.
+  - The waiting period is not necessary if a Lead approves the request.
+  - If an objection is raised, the request is put on hold until a Lead makes the final decision.
+- There is a maximum amount of issues with pledges someone can have assigned to them. The limit is 2 for core contributors, and 1 for anyone else. By limiting the amount of issues with pledges that someone can have assigned, we make sure the bounties remain available for others to pick up.
```
```suggestion
In some cases, Biome funds bounties on its own issues in order to further the project goals as laid out in its [roadmap](#roadmap). Because Biome is the one funding these issues, it also means **100% of the pledged amount will go to the contributor completing the task**. Project-Funded Bounties have the same conditions applied to them as [Community-Funded Bounties](#community-funded-bounties). But given that these issues are funded using money that ultimately comes from our sponsors, we have additional restrictions in place:

- Implementation of the task funded by Biome must advance our [roadmap](#roadmap) or help with essential infrastructure.
  - The task description must include a clear rationale for why this task is eligible for funding.
- To avoid paying a commission on our own bounties, we do not assign funds through Polar.sh. Instead, we ask contributors to take the money directly from our Open Collective account upon completion of the task.
  - Unfortunately, this means Project-Funded Bounties can only be assigned to [Core Contributors](#core-contributor) at this moment since they have access to our Open Collective account.

Additionally, there is a strict process for assigning Project-Funded Bounties:

- To request Biome funding for a task, a request with a clear rationale must be made in the `Community > #funding` channel on Discord.
  - To make sure our Core Contributors don't have to worry about overlooking Biome-funding-related decisions, requests made on GitHub issues are not sufficient for a funding request.
- If no Core Contributor raises objections within three (3) days, then any Core Contributor can assign the pledge to the issue and assign the requester to it.
  - The waiting period is optional if a Lead approves the request.
  - If an objection is raised, the request is put on hold until a Lead makes the final decision.
- There is a maximum amount of issues with pledges someone can have assigned to them. The limit is 2 for core contributors, and 1 for anyone else. By limiting the amount of issues with pledges that someone can have assigned, we make sure the bounties remain available for others to pick up.
```

Few grammar changes. The main suggestion is the change from 7 to 3. Mostly to align it with the voting period.

ematipico reviewed (approved):
Overall it looks good to me now!

nhedger reviewed (approved):
Thanks for taking the time to write this !

unvalley reviewed (approved):
Thank you for writing this up!
